lib/util/cache.js

```diff
@@ -0,0 +1,161 @@
+ *   's-maxage'?: number;
+ *   'stale-while-revalidate'?: number;
+ *   'stale-if-error'?: number;
+ *   public?: true;
+ *   private?: true;
+ *   'no-store'?: true;
+ *   'no-cache'?: true | string[];
+ *   'must-revalidate'?: true;
+ *   'proxy-revalidate'?: true;
+ *   immutable?: true;
+ *   'no-transform'?: true;
+ *   'must-understand'?: true;
+ *   'only-if-cached'?: true;
+ * }} CacheControlDirectives
+ *
+ * @param {string} header
+ * @returns {CacheControlDirectives}
+ */
+function parseCacheControlHeader (header) {
+  /**
+   * @type {import('../util/cache.js').CacheControlDirectives}
+   */
+  const output = {}
+
+  const directives = header.toLowerCase().split(',')
+  for (let i = 0; i < directives.length; i++) {
+    const directive = directives[i]
+    const keyValueDelimiter = directive.indexOf('=')
+
+    let key
+    let value
+    if (keyValueDelimiter !== -1) {
+      key = directive.substring(0, keyValueDelimiter).trim()
+      value = directive
+        .substring(keyValueDelimiter + 1, directive.length)
+        .trim()
+        .toLowerCase()
+    } else {
+      key = directive.trim()
+    }
+
+    switch (key) {
+      case 'min-fresh':
+      case 'max-stale':
+      case 'max-age':
+      case 's-maxage':
+      case 'stale-while-revalidate':
+      case 'stale-if-error': {
+        const parsedValue = parseInt(value, 10)
+        if (isNaN(parsedValue)) {
+          continue
+        }
+
+        output[key] = parsedValue
+
+        break
+      }
+      case 'private':
+      case 'no-cache': {
+        if (value) {
+          // The private and no-cache directives can be unqualified (aka just
+          //  `private` or `no-cache`) or qualified (w/ a value). When they're
+          //  qualified, it's a list of headers like `no-cache=header1`,
+          //  `no-cache="header1"`, or `no-cache="header1, header2"`
+          // If we're given multiple headers, the comma messes us up since
+          //  we split the full header by commas. So, let's loop through the
+          //  remaining parts in front of us until we find one that ends in a
+          //  quote. We can then just splice all of the parts in between the
+          //  starting quote and the ending quote out of the directives array
+          //  and continue parsing like normal.
+          // https://www.rfc-editor.org/rfc/rfc9111.html#name-no-cache-2
+          if (value[0] === '"') {
+            // Something like `no-cache="some-header"` OR `no-cache="some-header, another-header"`.
+
+            // Add the first header on and cut off the leading quote
+            const headers = [value.substring(1)]
+
+            let foundEndingQuote = false
+            for (let j = i; j < directives.length; j++) {
+              const nextPart = directives[j]
+              if (nextPart.endsWith('"')) {
+                foundEndingQuote = true
+                headers.push(...directives.splice(i + 1, j - 1).map(header => header.trim()))
```

Comment:
horrible! why??

Reply:
Since it's splitting the cache control header by commas, an entry like `no-cache="some-header, another-header"` messes it up. This finds the index of ` another-header` in the `directives` and splices from the index of `"some-header` + 1 to it.

Honestly the simplest way I could think of for doing this, not really happy with it either though. Went with the method of splitting the header by commas as opposed to going character by character just for the sake of simplicity as well.

Reply:
Can you please avoid all the splice and map and just use a for loop?